### ISSUE TITLE
Control over how Orleans behaves on CTRL+C

### ISSUE THREAD

**Issue by wattsm** [dotnet/orleans#3094]
The [Orleans Silo listens for CTRL+C](https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/Silo/Silo.cs#L526) and starts a fast exit when it detects it. This can cause problems if you use something like TopShelf to host a silo and are running it as a console application.

For example, if you configure TopShelf to gracefully shutdown the silo on exit:

```c#
service.WhenStopped(host => {

  host.ShutdownOrleansSilo();
  GC.SuppressFinalize(host);

});
```

When running as a Windows service this code will behave as you'd expect. However, if you run this as a console application then when you press CTRL+C to stop TopShelf Orleans will also respond to the event.

This causes intermittent exceptions on shutdown, which sometimes can leave the console application in a state which requires Task Manager / Visual Studio to kill it. I believe this is caused by the call to `host.ShutdownOrleansSilo()` happening after the silo begins its own fast exit process.

Example logs:

```
[2017-06-06 14:28:46.579 GMT    12	WARNING	100502	Catalog	127.0.0.1:11111]	UnregisterManyAsync 1 failed.	
Exc level 0: System.InvalidOperationException: Grain directory is stopping
   at Orleans.Runtime.GrainDirectory.LocalGrainDirectory.CheckIfShouldForward(GrainId grainId, Int32 hopCount, String operationDescription)
   at Orleans.Runtime.GrainDirectory.LocalGrainDirectory.UnregisterOrPutInForwardList(IEnumerable`1 addresses, UnregistrationCause cause, Int32 hopCount, Dictionary`2& forward, List`1 tasks, String context)
   at Orleans.Runtime.GrainDirectory.LocalGrainDirectory.<UnregisterManyAsync>d__108.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Orleans.Runtime.Scheduler.SchedulerExtensions.<>c__DisplayClass1_0.<<QueueTask>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Orleans.Runtime.Catalog.<FinishDestroyActivations>d__75.MoveNext()
```

Is there a way of detecting whether the silo is currently in the process of shutting down? I could not find one. The `SystemStatus` property of `Silo` would seem ideal, but is internal. Alternatively, can the silo configuration be extended to allow this behaviour to be switched off?

**Comment by benjaminpetit** [comment]
We don't currently expose an event when the silo stop.

I'm not sure we can remove this behavior (since maybe someone is relying in that), but I agree that we should extend the silo configuration to disable it.

**Comment by chris-eaton** [comment]
Just FYI I use TopShelf and havent encountered any problems with Ctrl+C in any of my apps [client or host]. Note though that I was using OrleansDashboard which stops orleans exiting on version 1.4.1. A new version was released today that fixes it

**Comment by sergeybykov** [comment]
Resolved via #3109.

**Comment by bboyle1234** [comment]
@wattsm  see https://github.com/dotnet/orleans/issues/4151 for a possible fix
